src/python/pants/engine/target.py

```diff
@@ -305,6 +305,18 @@ class PluginField:
         plugins can use that Field like any other Field.
         """
 
+        # NB: It's important that every subclass of `Target` have its own distinct `PluginField`
+        # type so that registering a PluginField for one target doesn't end up registering for all
+        # targets. This means that we cannot use a statically declared class, so we instead
+        # dynamically create a class. We use a class property and name the method in PascalCase so
+        # that this implementation doesn't leak to call sites.
```

Comment:
I'd prefer just a `@memoized_classmethod` and not pretending this is a nested type. Saying `SomeTarget.plugin_field()` is ~no more typing and less magic. Maybe better `SomeTarget.register_plugin_field(Field) -> UnionRule` - absolutely no magic.

Reply:
`SomeTarget.register_plugin_field(Field) -> UnionRule` I like that.

Reply:
I'm fine either way. It would be best if that ended up being a pattern we used elsewhere as well... ie, `Subsystem.register() -> TaskRule` or something...